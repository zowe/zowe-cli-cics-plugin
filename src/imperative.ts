/*
* This program and the accompanying materials are made available under the terms of the *
* Eclipse Public License v2.0 which accompanies this distribution, and is available at *
* https://www.eclipse.org/legal/epl-v20.html                                      *
*                                                                                 *
* SPDX-License-Identifier: EPL-2.0                                                *
*                                                                                 *
* Copyright Contributors to the Zowe Project.                                     *
*                                                                                 *
*/

// Imperative version of Zowe CLI
import { IImperativeConfig } from "@brightside/imperative";
import { PluginConstants } from "./api/constants/PluginConstants";
import { CicsSession } from "./cli/CicsSession";

const config: IImperativeConfig = {
    commandModuleGlobs: ["**/cli/*/*.definition!(.d).*s"],
    rootCommandDescription: PluginConstants.PLUGIN_DESCRIPTION,
    productDisplayName: PluginConstants.PLUGIN_NAME,
    name: PluginConstants.PLUGIN_GROUP_NAME,
    pluginHealthCheck: "./lib/healthCheck.handler",
    profiles: [
        {
            type: "cics",
            schema: {
                type: "object",
                title: "CICS Profile",
                description: "A cics profile is required to issue commands in the cics command group that interact with " +
                    "CICS regions. The cics profile contains your host, port, user name, and password " +
                    "for the IBM CICS management client interface (CMCI) server of your choice.",
                properties: {
                    host: {
                        type: "string",
                        optionDefinition: {
                            name: "host",
                            aliases: ["H"],
                            description: "The CMCI server host name",
                            type: "string",
                            required: true,
                        },
                    },
                    port: {
                        type: "number",
                        optionDefinition: {
                            name: "port",
                            aliases: ["P"],
                            description: "The CMCI server port",
                            type: "number",
                            defaultValue: 1490,
                        },
                    },
                    user: {
                        type: "string",
                        secure: true,
                        optionDefinition: {
                            name: "user",
                            aliases: ["u"],
                            description: "Your username to connect to CICS",
                            type: "string",
                            implies: ["password"],
                            required: true,
                        },
                    },
                    password: {
                        type: "string",
                        secure: true,
                        optionDefinition: {
                            name: "password",
                            aliases: ["p"],
                            description: "Your password to connect to CICS",
                            type: "string",
                            implies: ["user"],
                            required: true,
                        },
                    },
                    regionName: {
                        type: "string",
                        optionDefinition: {
                            name: "region-name",
                            description: "The name of the CICS region name to interact with",
                            type: "string"
                        },
                    },
                    cicsPlex: {
                        type: "string",
                        optionDefinition: {
                            name: "cics-plex",
                            description: "The name of the CICSPlex to interact with",
                            type: "string"
                        },
                    },
                    rejectUnauthorized: {
                        type: "boolean",
                        optionDefinition: CicsSession.CICS_OPTION_REJECT_UNAUTHORIZED
                    },
                    protocol: {
                        type: "string",
                        optionDefinition: CicsSession.CICS_OPTION_PROTOCOL
                    }
                },
                required: ["host"],
            },
            createProfileExamples: [
                {
                    options: "cics123 --host zos123 --port 1490 --user ibmuser --password myp4ss",
                    description: "Create a cics profile named 'cics123' to connect to CICS at host zos123 and port 1490"
                }
            ]
        }
    ]
};

export = config;
