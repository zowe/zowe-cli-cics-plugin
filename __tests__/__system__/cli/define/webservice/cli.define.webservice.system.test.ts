/*
* This program and the accompanying materials are made available under the terms of the *
* Eclipse Public License v2.0 which accompanies this distribution, and is available at *
* https://www.eclipse.org/legal/epl-v20.html                                      *
*                                                                                 *
* SPDX-License-Identifier: EPL-2.0                                                *
*                                                                                 *
* Copyright Contributors to the Zowe Project.                                     *
*                                                                                 *
*/

import { ITestEnvironment, TestEnvironment, runCliScript } from "@zowe/cli-test-utils";
import { ITestPropertiesSchema } from "../../../../__src__/doc/ITestPropertiesSchema";
import { generateRandomAlphaNumericString } from "../../../../__src__/TestUtils";
import { Session } from "@zowe/imperative";
import { IWebServiceParms } from "../../../../../src";
import { deleteWebservice } from "../../../../../src/api/methods/delete/Delete";

let TEST_ENVIRONMENT: ITestEnvironment<ITestPropertiesSchema>;
let regionName: string;
let csdGroup: string;
let host: string;
let port: number;
let user: string;
let password: string;
let protocol: string;
let rejectUnauthorized: boolean;
let session: Session;

describe("CICS define web service command", () => {

    beforeAll(async () => {
        TEST_ENVIRONMENT = await TestEnvironment.setUp({
            testName: "define_webservice",
            installPlugin: true,
            tempProfileTypes: ["cics"]
        });
        const cicsProperties = TEST_ENVIRONMENT.systemTestProperties.cics;
        csdGroup = TEST_ENVIRONMENT.systemTestProperties.cmci.csdGroup;
        regionName = TEST_ENVIRONMENT.systemTestProperties.cmci.regionName;
        host = TEST_ENVIRONMENT.systemTestProperties.cics.host;
        port = TEST_ENVIRONMENT.systemTestProperties.cics.port;
        user = TEST_ENVIRONMENT.systemTestProperties.cics.user;
        password = TEST_ENVIRONMENT.systemTestProperties.cics.password;
        protocol = TEST_ENVIRONMENT.systemTestProperties.cics.protocol;
        rejectUnauthorized = TEST_ENVIRONMENT.systemTestProperties.cics.rejectUnauthorized;
        session = new Session({
            type: "basic",
            hostname: cicsProperties.host,
            port: cicsProperties.port,
            user: cicsProperties.user,
            password: cicsProperties.password,
            rejectUnauthorized: cicsProperties.rejectUnauthorized || false,
            protocol: cicsProperties.protocol as any || "https",
        });
    });

    afterAll(async () => {
        await TestEnvironment.cleanUp(TEST_ENVIRONMENT);
    });

    it("should be able to display the help", () => {
        const output = runCliScript(__dirname + "/__scripts__/define_webservice_help.sh", TEST_ENVIRONMENT, []);
        expect(output.stderr.toString()).toEqual("");
        expect(output.status).toEqual(0);
        expect(output.stdout.toString()).toMatchSnapshot();
    });

    it("should be able to successfully define a web service with basic options", async () => {
        const websvcNameSuffixLength = 4;
        const websvcName = "DFN" + generateRandomAlphaNumericString(websvcNameSuffixLength);
        const options: IWebServiceParms = { name: websvcName, csdGroup, regionName };
        let output = runCliScript(__dirname + "/__scripts__/define_webservice.sh", TEST_ENVIRONMENT,
            [websvcName, csdGroup, "FAKEPIPE", "//u/exampleapp/wsbind/example.log", regionName]);
        let stderr = output.stderr.toString();
        expect(stderr).toEqual("");
        expect(output.status).toEqual(0);
        expect(output.stdout.toString()).toContain("success");

        output = runCliScript(__dirname + "/__scripts__/get_resource_webservice.sh", TEST_ENVIRONMENT,
            [regionName, csdGroup]);
        stderr = output.stderr.toString();
        expect(stderr).toEqual("");
        expect(output.status).toEqual(0);
        expect(output.stdout.toString()).toContain(websvcName);

        await deleteWebservice(session, options);
    });

    it("should get a syntax error if web service name is omitted", () => {
        const output = runCliScript(__dirname + "/__scripts__/define_webservice.sh", TEST_ENVIRONMENT,
            ["", "FAKEGRP", "FAKEPIPE", "FAKEWSB", "FAKERGN"]);
        const stderr = output.stderr.toString();
        expect(stderr).toContain("Syntax");
        expect(stderr).toContain("webserviceName");
        expect(output.status).toEqual(1);
    });

    it("should get a syntax error if CSD group is omitted", () => {
        const output = runCliScript(__dirname + "/__scripts__/define_webservice.sh", TEST_ENVIRONMENT,
            ["FAKEWSVC", "", "FAKEPIPE", "FAKEWSB", "FAKERGN"]);
        const stderr = output.stderr.toString();
        expect(stderr).toContain("Syntax");
        expect(stderr).toContain("csdGroup");
        expect(output.status).toEqual(1);
    });

    it("should get a syntax error if web service pipeline name is omitted", () => {
        const output = runCliScript(__dirname + "/__scripts__/define_webservice.sh", TEST_ENVIRONMENT,
            ["FAKEWSVC", "FAKEGRP", "", "FAKEWSB", "FAKERGN"]);
        const stderr = output.stderr.toString();
        expect(stderr).toContain("Syntax");
        expect(stderr).toContain("pipeline-name");
        expect(output.status).toEqual(1);
    });

    it("should get a syntax error if web service binding file is omitted", () => {
        const output = runCliScript(__dirname + "/__scripts__/define_webservice.sh", TEST_ENVIRONMENT,
            ["FAKEWSVC", "FAKEGRP", "FAKEPIPE", "", "FAKERGN"]);
        const stderr = output.stderr.toString();
        expect(stderr).toContain("Syntax");
        expect(stderr).toContain("wsbind");
        expect(output.status).toEqual(1);
    });

    it("should get a syntax error if region name is omitted", () => {
        const output = runCliScript(__dirname + "/__scripts__/define_webservice.sh", TEST_ENVIRONMENT,
            ["FAKEWSVC", "FAKEGRP", "FAKEPIPE", "FAKEWSB", ""]);
        const stderr = output.stderr.toString();
        expect(stderr).toContain("Syntax");
        expect(stderr).toContain("region-name");
        expect(output.status).toEqual(1);
    });

    it("should be able to successfully define a web service using profile options", async () => {
        const websvcNameSuffixLength = 4;
        const websvcName = "DFN" + generateRandomAlphaNumericString(websvcNameSuffixLength);
        const pipelineName = "FAKEPIPE";
        const wsBind = "//u/exampleapp/wsbind/example.log";
        const options: IWebServiceParms = { name: websvcName, csdGroup, regionName };
        let output = runCliScript(__dirname + "/__scripts__/define_webservice_fully_qualified.sh", TEST_ENVIRONMENT,
            [websvcName,
                csdGroup,
                pipelineName,
                wsBind,
                regionName,
                host,
                port,
                user,
                password,
                protocol,
                rejectUnauthorized]);
        let stderr = output.stderr.toString();
        expect(stderr).toEqual("");
        expect(output.status).toEqual(0);
        expect(output.stdout.toString()).toContain("success");

        output = runCliScript(__dirname + "/__scripts__/get_resource_webservice_fully_qualified.sh", TEST_ENVIRONMENT,
            [regionName,
                csdGroup,
                host,
                port,
                user,
                password,
                protocol,
                rejectUnauthorized]);
        stderr = output.stderr.toString();
        expect(stderr).toEqual("");
        expect(output.status).toEqual(0);
        expect(output.stdout.toString()).toContain(websvcName);

        await deleteWebservice(session, options);
    });

});
