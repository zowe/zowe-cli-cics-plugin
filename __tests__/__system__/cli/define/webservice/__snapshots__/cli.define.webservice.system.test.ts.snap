// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CICS define web service command should be able to display the help 1`] = `
"
 COMMAND NAME
 ------------

   webservice | web

 DESCRIPTION
 -----------

   Define a new web service to CICS.

 USAGE
 -----

   zowe cics define webservice <webserviceName> <csdGroup> [options]

 POSITIONAL ARGUMENTS
 --------------------

   webserviceName		 (string)

      The name of the WEBSERVICE to create. The maximum length of the web service name
      is eight characters.

   csdGroup		 (string)

      The CICS system definition (CSD) Group for the new web service that you want to
      define. The maximum length of the group name is eight characters.

 REQUIRED OPTIONS
 ----------------

   --pipeline-name | --pn (string)

      The name of the PIPELINE resource definition for the web service. The maximum
      length of the pipeline name is eight characters

   --wsbind (string)

      The file name of the web service binding file on HFS.

 OPTIONS
 -------

   --description | --desc (string)

      Description of the web service resource being defined.

   --validation (boolean)

      Specifies whether full validation of SOAP messages against the corresponding
      schema in the web service description should be performed at run time.

      Default value: false

   --wsdlfile | --wsdl (string)

      The file name of the web service description (WSDL) file on HFS.

   --region-name (string)

      The CICS region name to which to define the new web service.

   --cics-plex (string)

      The name of the CICSPlex to which to define the new web service.

 CICS CONNECTION OPTIONS
 -----------------------

   --host | -H (string)

      The CICS server host name.

   --port | -P (number)

      The CICS server port.

      Default value: 443

   --user | -u (string)

      Mainframe (CICS) user name, which can be the same as your TSO login.

   --password | --pw (string)

      Mainframe (CICS) password, which can be the same as your TSO password.

   --reject-unauthorized | --ru (boolean)

      Reject self-signed certificates.

      Default value: true

   --protocol | -o (string)

      Specifies CMCI protocol (http or https).

      Default value: https
      Allowed values: http, https

 PROFILE OPTIONS
 ---------------

   --cics-profile | --cics-p (string)

      The name of a (cics) profile to load for this command execution.

   --base-profile | --base-p (string)

      The name of a (base) profile to load for this command execution.

 BASE CONNECTION OPTIONS
 -----------------------

   --token-type | --tt (string)

      The type of token to get and use for the API. Omit this option to use the
      default token type, which is provided by 'zowe auth login'.

   --token-value | --tv (string)

      The value of the token to pass to the API.

   --cert-file (local file path)

      The file path to a certificate file to use for authentication

   --cert-key-file (local file path)

      The file path to a certificate key file to use for authentication

 GLOBAL OPTIONS
 --------------

   --response-format-json | --rfj (boolean)

      Produce JSON formatted data from a command

   --help | -h (boolean)

      Display help text

   --help-examples (boolean)

      Display examples for all the commands in a group

   --help-web | --hw (boolean)

      Display HTML help in browser

 EXAMPLES
 --------

   - Define a webservice named WEBSVCA for the pipeline named
   PIPE123 to the region named MYREGION in the CSD group MYGRP where the binding
   file is /u/exampleapp/wsbind/example.log:

      $ zowe cics define webservice WEBSVCA MYGRP --pipeline-name PIPELINE --wsbind /u/exampleapp/wsbind/example.log --region-name MYREGION

"
`;
